### DISCUSSION TITLE
Incorrect DDL for composite pk with autoincrement column and Incorrect foreign keys constraints to composite pk

### DISCUSSION THREAD

**Discussion by petruconstantin** [mikro-orm/mikro-orm#3137] [Q&A]
**Describe the bug**

When using foreign keys to composite primary key with autoincrement id the table definition and queries are not generated correctly
The produced table definition contains one column for each component of the composite foreign key e.g. 1 column for project_id and project_tenant_id

**To Reproduce**

using the following tables
```
@Entity({ tableName: 'tenant' })
export class Tenant {
    @PrimaryKey()
    id!: number

    @Property({ length: 5, nullable: false, unique: true })
    @Index()
    @IsString({})
    slug!: string

    @Property({ nullable: false, unique: true })
    name!: string
}


@Entity({ abstract: true })
export abstract class TenantEntity {

    @PrimaryKey({ autoincrement: true, unique: true })
    id!: number

    @ManyToOne({
        hidden: true,
        nullable: false,
        entity: () => Tenant,
        eager: false,
        primary: true
    })
    tenant?: Tenant
}

@Entity({ tableName: 'project' })
export class Project extends   TenantEntity  {
    @Property({ length: 255, nullable: true })
    name?: string
}

Entity({ tableName: 'test' })
export class Test extends TenantEntity {
    @ManyToOne({
        entity: () => UlpProject,
        nullable: true
    })
    project?: UlpProject
}

**Expected behavior**

the expected ddl output:
```

CREATE TABLE `test` (
  `id` int unsigned NOT NULL AUTO_INCREMENT,
  `tenant` int unsigned NOT NULL,
  `project_id` int unsigned DEFAULT NULL,
  PRIMARY KEY (`id`,`tenant`),
  KEY `test_FK` (`project_id`,`tenant`),
  CONSTRAINT `test_FK` FOREIGN KEY (`project_id`, `tenant`) REFERENCES `project` (`id`, `tenant_id`)
) ;
```


**Additional context**
Add any other context about the problem here.

**Versions**

| Dependency | Version |
| - | - |
| node | v16.15.0 |
| typescript | 4.6 |
| mikro-orm | 5.1.4 |
| your-driver | mysql |

**Comment by B4nan** [comment]
You get what you define in your entities. A FK to composite PK is always defined by all of its parts. Your `Test` entity has 2 columns describing its PK and 2 columns describing the FK. Both the PK and FK have two parts. Not sure why you expect the tenant property should be somehow automatically shared between them - it should not.

I would suggest removing the tenant from PK, autoincrement column on its own is enough for a PK and you can have the tenant column defined next to it. Having composite PK that includes autoincrement column on its own is not supported, you would have issues saving such entities.

**Reply by petruconstantin** [reply]
Thanks for taking time to analyze this.

Below I pasted a small db example with the behavior I'm trying to replicate. 
If the tenant_id is dropped from pk then the DB constraints would not forbid inserting the phase on a tenant with a project from another tenant. 

Given the id is guaranteed to be unique and autoincremented on its own it can be referenced in a foreign key.

Please let me know what's your take on this approach.

```
DROP TABLE IF EXISTS `tenant`;
CREATE TABLE `tenant` (
  `id` int unsigned NOT NULL AUTO_INCREMENT,
  `name` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=3 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;

DROP TABLE IF EXISTS `project`;
CREATE TABLE `project` (
  `id` int unsigned NOT NULL AUTO_INCREMENT,
  `tenant_id` int unsigned NOT NULL,
  `name` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`tenant_id`,`id`),
   UNIQUE KEY `project_un` (`id`),
  CONSTRAINT `project_FK_1` FOREIGN KEY (`tenant_id`) REFERENCES `tenant` (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=4 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;

DROP TABLE IF EXISTS `phase`;
CREATE TABLE `phase` (
  `id` int unsigned NOT NULL AUTO_INCREMENT,
  `tenant_id` int unsigned NOT NULL,
  `project_id` int unsigned DEFAULT NULL,
  PRIMARY KEY (`id`,`tenant_id`),
  UNIQUE KEY `phase_un` (`id`),
  KEY `phase_FK` (`tenant_id`,`id`),
  KEY `phase_FK_3` (`tenant_id`,`project_id`),
  CONSTRAINT `phase_FK` FOREIGN KEY (`tenant_id`) REFERENCES `tenant` (`id`),
  CONSTRAINT `phase_FK_3` FOREIGN KEY (`tenant_id`, `project_id`) REFERENCES `project` (`tenant_id`, `id`)
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;


INSERT INTO `tenant` 	VALUES (1,'ENTITY 1'),(2,'ENTITY 2');
INSERT INTO `project` 	VALUES (1,1,'project 1'),(3,1,'project 3'),(2,2,'project 2');

-- this should pass
INSERT INTO `phase` 	VALUES (1,2,2);

-- this should fail
INSERT INTO `phase` 	VALUES (1,2,2);
```

**Reply by B4nan** [reply]
I do understand what you are up to, but its not supported, neither on schema level, nor on runtime level. Do you know how other ORMs handle this? Especially the entity definition bit.

Btw looked at hibernate docs, and I very much related to what they say about the mixed PKs with part being generated:

<img width="690" alt="image" src="https://user-images.githubusercontent.com/615580/169808595-cef6762c-ba6a-4a44-abcc-bb3067dcd1bc.png">

**Reply by petruconstantin** [reply]
Then I guess it's not a feasible solution :( 
I'll look a bit into EF but I guess I'll modify my architecture to multiple-schemas.

Thank you for your response and for the awesome lib - it's really great :)
